Bearded.Graphics/Debugging/KHRDebugExtension.cs

```diff
@@ -0,0 +1,57 @@
+using System.Linq;
+using OpenTK.Graphics.OpenGL;
+
+namespace Bearded.Graphics.Debugging
+{
+    public abstract class KHRDebugExtension
+    {
+        private static KHRDebugExtension? instance;
+        public static KHRDebugExtension Instance => instance ??= createInstanceWithAvailableFunctionality();
+
+        private KHRDebugExtension()
+        {
+        }
+
+        private static KHRDebugExtension createInstanceWithAvailableFunctionality()
+        {
+            if (khrDebugExtensionIsAvailable())
+                return new ActualImplementation();
+            return new DummyImplementation();
+        }
+
+        private static bool khrDebugExtensionIsAvailable()
+        {
+            var count = GL.GetInteger(GetPName.NumExtensions);
+
+            return Enumerable.Range(0, count)
+                .Select(i => GL.GetString(StringNameIndexed.Extensions, i))
+                .Contains("GL_KHR_debug");
+        }
+
+        public virtual void SetObjectLabel(ObjectLabelIdentifier identifier, int name, string label) { }
+        public virtual void PushDebugGroup(DebugSourceExternal source, int id, string name) { }
+        public virtual void PopDebugGroup() { }
+
+        private sealed class ActualImplementation : KHRDebugExtension
+        {
+            public override void SetObjectLabel(ObjectLabelIdentifier identifier, int name, string label)
+            {
+                GL.ObjectLabel(identifier, name, label.Length, label);
+            }
+
+            public override void PushDebugGroup(DebugSourceExternal source, int id, string name)
```

Comment:
For my curiosity: what is the id here used for? It seems to always be hardcoded to 0, so does it make sense exposing this at all?

Reply:
Documentation says:
> The id​ is purely for the benefit of the user and may be any unsigned 32-bit integer value.

Basically, if I understand correctly, it's for you to use for whatever purposes you consider important, if you rather not match by string.

This class specifically isn't trying to simplify the API except removing (mostly) redundant stuff, it's just here to expose the extension in a way that's agnostic to whether it actually exists.

The pipeline layer is where we abstract all of that away then and you only pass the string. To me this seems the right way of doing it.